# EDA: Diagnosing Diabetes

In this project, you'll imagine you are a data scientist interested in exploring data that looks at how certain diagnostic factors affect the diabetes outcome of women patients.

You will use your EDA skills to help inspect, clean, and validate the data.

**Note**: This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database) is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration per 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure
- `SkinThickness`: Triceps skinfold thickness
- `Insulin`: 2-Hour serum insulin
- `BMI`: Body mass index
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

Let's get started!

## Initial Inspection

1. First, familiarize yourself with the dataset [here](https://www.kaggle.com/uciml/pima-indians-diabetes-database).

   Look at each of the nine columns in the documentation.
   
   What do you expect each data type to be?

Expected data type for each column:

- `Pregnancies`: int
- `Glucose`: float
- `BloodPressure`: float
- `SkinThickness`: int
- `Insulin`: float
- `BMI`: float
- `DiabetesPedigreeFunction`: float
- `Age`: int
- `Outcome`: bool

2. Next, let's load in the diabetes data to start exploring.

   Load the data in a variable called `diabetes_data` and print the first few rows.
   
   **Note**: The data is stored in a file called `diabetes.csv`.

In [1]:
import pandas as pd
import numpy as np

# load in data
diabetes_data = pd.read_csv('diabetes.csv')

print(diabetes_data.head())


   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1  


3. How many columns (features) does the data contain?

In [2]:
# print number of columns
print(len(diabetes_data.columns))

9


4. How many rows (observations) does the data contain?

In [3]:
# print number of rows
#we can just use the count() function on any column

print(diabetes_data.Pregnancies.count())

768


## Further Inspection

5. Let's inspect `diabetes_data` further.

   Do any of the columns in the data contain null (missing) values?

In [4]:
# find whether columns contain null values

diabetes_data_null = diabetes_data[diabetes_data.isnull().any(axis=1)]

print(diabetes_data_null)


Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []


6. If you answered no to the question above, not so fast!

   While it's technically true that none of the columns contain null values, that doesn't necessarily mean that the data isn't missing any values.
   
   When exploring data, you should always question your assumptions and try to dig deeper.
   
   To investigate further, calculate summary statistics on `diabetes_data` using the `.describe()` method.

In [5]:
# perform summary statistics
print(diabetes_data.describe())

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000        

7. Looking at the summary statistics, do you notice anything odd about the following columns?

   - `Glucose`
   - `BloodPressure`
   - `SkinThickness`
   - `Insulin`
   - `BMI`

**Your response to question 7**: It's not possible in practice to find a value of 0 for these variables. It indicates non-measured values set to 0.

8. Do you spot any other outliers in the data?

**Your response to question 8**: 
Pregnancies max of 17 while average is around 4
Skin thickness max of 99, 5 times the average
Insulin max of 846, more than 10* the average
Age max of 81 yo, more than 2 times the average


9. Let's see if we can get a more accurate view of the missing values in the data.

   Use the following code to replace the instances of `0` with `NaN` in the five columns mentioned:
   
   ```py
   diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)
   ```

In [6]:
# replace instances of 0 with NaN
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)


10. Next, check for missing (null) values in all of the columns just like you did in Step 5.

    Now how many missing values are there?

In [7]:
# find whether columns contain null values after replacements are made
diabetes_data_null = diabetes_data[diabetes_data.isnull().any(axis=1)]
print(len(diabetes_data_null))



376


11. Let's take a closer look at these rows to get a better idea of _why_ some data might be missing.

    Print out all the rows that contain missing (null) values.

In [12]:
# print rows with missing values

print(diabetes_data_null)


for i in range(1,6):
    col = diabetes_data_null.columns[i]
    null = diabetes_data_null[col].isnull()
    print(f'The column {col} has {len(diabetes_data_null[null])} missing values')
    





     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
5              5    116.0           74.0            NaN      NaN  25.6   
7             10    115.0            NaN            NaN      NaN  35.3   
..           ...      ...            ...            ...      ...   ...   
761            9    170.0           74.0           31.0      NaN  44.0   
762            9     89.0           62.0            NaN      NaN  22.5   
764            2    122.0           70.0           27.0      NaN  36.8   
766            1    126.0           60.0            NaN      NaN  30.1   
767            1     93.0           70.0           31.0      NaN  30.4   

     DiabetesPedigreeFunction  Age Outcome  
0                       0.627   50       1  
1                    

12. Go through the rows with missing data. Do you notice any patterns or overlaps between the missing data?

**Your response to question 12**: We observe that in almost all cases (374 out of 376 rows), the Insulin value is missing. The next most missing value is Skin Thickness. As for the rest of the variables, the number of missing values is relatively small. 

13. Next, take a closer look at the data types of each column in `diabetes_data`.

    Does the result match what you would expect?

In [9]:
# print data types using .info() method
print(diabetes_data.dtypes)


Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                      object
dtype: object


14. To figure out why the `Outcome` column is of type `object` (string) instead of type `int64`, print out the unique values in the `Outcome` column.

In [10]:
# print unique values of Outcome column
print(diabetes_data.Outcome.unique())


['1' '0' 'O']


15. How might you resolve this issue?

**Your response to question 15**: Just by replacing the "O" values by 0. It is done in the below cell.

In [11]:
diabetes_data['Outcome'] = diabetes_data['Outcome'].replace('O', '0')
print(diabetes_data.Outcome.unique())


['1' '0']


## Next Steps:

16. Congratulations! In this project, you saw how EDA can help with the initial data inspection and cleaning process. This is an important step as it helps to keep your datasets clean and reliable.

    Here are some ways you might extend this project if you'd like:
    - Use `.value_counts()` to more fully explore the values in each column.
    - Investigate other outliers in the data that may be easily overlooked.
    - Instead of changing the `0` values in the five columns to `NaN`, try replacing the values with the median or mean of each column.

In [57]:
# VALUE COUNTS

#Testing the value counts function on a few columns
print(diabetes_data.Outcome.value_counts())
print(diabetes_data.Age.value_counts())
print(diabetes_data.Pregnancies.value_counts())

#Applying the function to a column featuring a continuous variable (BMI) yields much more results
print(diabetes_data.BMI.value_counts())

# OUTLIERS

#Can we find some outliers in the 'Pregnancies' columns, with exceptionnally high number of pregnancies?
#First calculate the mean: a bit less than 4
print(diabetes_data.Pregnancies.mean())
print(diabetes_data[diabetes_data.Pregnancies >= 8])
#124 patients with at least two times more pregnancies than the average

#Can we find outliers, up and low, in the 'BMI' column?
mean = diabetes_data.BMI.mean()
std = diabetes_data.BMI.std()
print(diabetes_data.BMI.mean())
print(diabetes_data.BMI.std())
#Average is approx 34 and std deviation approx 7. Can we find values that are distant of at least 2 std deviations from the mean?
print(diabetes_data[(diabetes_data.BMI >= mean +2*std) | (diabetes_data.BMI <= mean - 2*std)])
print(len(diabetes_data[(diabetes_data.BMI >= mean +2*std) | (diabetes_data.BMI <= mean - 2*std)]))
print(len(diabetes_data[diabetes_data.BMI >= mean +2*std]))
#24 values are concerned, the most of them (20) being upper outliers

#REPLACE BY MEAN instead of NaN

diabetes_data = pd.read_csv('diabetes.csv')
columns_to_change = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']



for col in columns_to_change:    
    #we make sure to give all missing values the same number, initial mean, instead of recalculating the mean each time
    set = diabetes_data[diabetes_data[col] > 0]    
    mean_t = set[col].mean()
    diabetes_data[col] = diabetes_data[col].replace(0,mean_t)





Outcome
0    494
1    268
O      6
Name: count, dtype: int64
Age
22    72
21    63
25    48
24    46
23    38
28    35
26    33
27    32
29    29
31    24
41    22
30    21
37    19
42    18
33    17
38    16
36    16
32    16
45    15
34    14
46    13
43    13
40    13
39    12
35    10
50     8
51     8
52     8
44     8
58     7
47     6
54     6
49     5
48     5
57     5
53     5
60     5
66     4
63     4
62     4
55     4
67     3
56     3
59     3
65     3
69     2
61     2
72     1
81     1
64     1
70     1
68     1
Name: count, dtype: int64
Pregnancies
1     135
0     111
2     103
3      75
4      68
5      57
6      50
7      45
8      38
9      28
10     24
11     11
13     10
12      9
14      2
15      1
17      1
Name: count, dtype: int64
BMI
32.000000    13
31.600000    12
31.200000    12
32.457464    11
32.400000    10
             ..
36.700000     1
41.800000     1
42.600000     1
42.800000     1
46.300000     1
Name: count, Length: 248, dtype: int64
3.845052083333